File: ocaml/test/test_parser.ml
let () =
  let open Alcotest in
  let open Ccc_lib in
  let pa ~runner ~thinking ~provider ~model ~alias ~prompt =
      { Parser.runner; Parser.thinking; Parser.provider;
        Parser.model; Parser.alias; Parser.prompt }
  in
  run "parser" [
    ("parse_args", [
      test_case "prompt only" `Quick (fun () ->
        let r = Parser.parse_args ["hello"] in
        check (option string) "runner" None r.Parser.runner;
        check (option int) "thinking" None r.Parser.thinking;
        check (option string) "provider" None r.Parser.provider;
        check (option string) "model" None r.Parser.model;
        check (option string) "alias" None r.Parser.alias;
        check string "prompt" "hello" r.Parser.prompt);

      test_case "multi-word prompt" `Quick (fun () ->
        let r = Parser.parse_args ["hello"; "world"] in
        check string "prompt" "hello world" r.Parser.prompt;
        check (option string) "runner" None r.Parser.runner);

      test_case "runner selector cc" `Quick (fun () ->
        let r = Parser.parse_args ["cc"; "test"] in
        check (option string) "runner" (Some "cc") r.Parser.runner;
        check string "prompt" "test" r.Parser.prompt);

      test_case "runner selector c" `Quick (fun () ->
        let r = Parser.parse_args ["c"; "test"] in
        check (option string) "runner" (Some "c") r.Parser.runner;
        check string "prompt" "test" r.Parser.prompt);

      test_case "runner selector cx" `Quick (fun () ->
        let r = Parser.parse_args ["cx"; "test"] in
        check (option string) "runner" (Some "cx") r.Parser.runner;
        check string "prompt" "test" r.Parser.prompt);

      test_case "runner selector case insensitive" `Quick (fun () ->
        let r = Parser.parse_args ["CC"; "test"] in
        check (option string) "runner" (Some "cc") r.Parser.runner);

      test_case "thinking" `Quick (fun () ->
        let r = Parser.parse_args ["+2"; "test"] in
        check (option int) "thinking" (Some 2) r.Parser.thinking;
        check string "prompt" "test" r.Parser.prompt);

      test_case "provider:model" `Quick (fun () ->
        let r = Parser.parse_args [":anthropic:sonnet-4"; "test"] in
        check (option string) "provider" (Some "anthropic") r.Parser.provider;
        check (option string) "model" (Some "sonnet-4") r.Parser.model;
        check string "prompt" "test" r.Parser.prompt);

      test_case "model only" `Quick (fun () ->
        let r = Parser.parse_args [":sonnet-4"; "test"] in
        check (option string) "model" (Some "sonnet-4") r.Parser.model;
        check (option string) "provider" None r.Parser.provider);

      test_case "alias" `Quick (fun () ->
        let r = Parser.parse_args ["@review"; "test"] in
        check (option string) "alias" (Some "review") r.Parser.alias;
        check string "prompt" "test" r.Parser.prompt);

      test_case "full combo" `Quick (fun () ->
        let r = Parser.parse_args
          ["claude"; "+3"; ":anthropic:sonnet-4"; "@review"; "fix"; "bugs"] in
        check (option string) "runner" (Some "claude") r.Parser.runner;
        check (option int) "thinking" (Some 3) r.Parser.thinking;
        check (option string) "provider" (Some "anthropic") r.Parser.provider;
        check (option string) "model" (Some "sonnet-4") r.Parser.model;
        check (option string) "alias" (Some "review") r.Parser.alias;
        check string "prompt" "fix bugs" r.Parser.prompt);

      test_case "positional swallows flags" `Quick (fun () ->
        let r = Parser.parse_args ["hello"; "+2"] in
        check string "prompt" "hello +2" r.Parser.prompt;
        check (option int) "thinking" None r.Parser.thinking);

      test_case "empty argv" `Quick (fun () ->
        let r = Parser.parse_args [] in
        check string "prompt" "" r.Parser.prompt);

      test_case "second runner becomes positional" `Quick (fun () ->
        let r = Parser.parse_args ["claude"; "oc"; "test"] in
        check (option string) "runner" (Some "claude") r.Parser.runner;
        check string "prompt" "oc test" r.Parser.prompt);

      test_case "abbrev runner oc" `Quick (fun () ->
        let r = Parser.parse_args ["oc"; "test"] in
        check (option string) "runner" (Some "oc") r.Parser.runner);
    ]);

    ("resolve_command", [
      test_case "default runner opencode" `Quick (fun () ->
        let parsed = Parser.parse_args ["hello"] in
        let argv, env, _warnings = Parser.resolve_command parsed None in
        check (list string) "argv" ["opencode"; "run"; "hello"] argv;
        check (list (pair string string)) "env" [] env);

      test_case "claude runner" `Quick (fun () ->
        let parsed = Parser.parse_args ["claude"; "hello"] in
        let argv, env, _warnings = Parser.resolve_command parsed None in
        check (list string) "argv" ["claude"; "hello"] argv;
        check (list (pair string string)) "env" [] env);

      test_case "claude runner via selector cc" `Quick (fun () ->
        let parsed = Parser.parse_args ["cc"; "hello"] in
        let argv, env, _warnings = Parser.resolve_command parsed None in
        check (list string) "argv" ["claude"; "hello"] argv;
        check (list (pair string string)) "env" [] env);

      test_case "claude thinking +3" `Quick (fun () ->
        let parsed = Parser.parse_args ["claude"; "+3"; "hello"] in
        let argv, env, _warnings = Parser.resolve_command parsed None in
        check (list string) "argv"
          ["claude"; "--thinking"; "enabled"; "--effort"; "high"; "hello"] argv;
        check (list (pair string string)) "env" [] env);

      test_case "claude thinking 0 (no-thinking)" `Quick (fun () ->
        let parsed = Parser.parse_args ["claude"; "+0"; "hello"] in
        let argv, _, _warnings = Parser.resolve_command parsed None in
        check (list string) "argv" ["claude"; "--thinking"; "disabled"; "hello"] argv);

      test_case "codex model flag" `Quick (fun () ->
        let parsed = Parser.parse_args ["codex"; ":gpt-4"; "hello"] in
        let argv, _, _warnings = Parser.resolve_command parsed None in
        check (list string) "argv" ["codex"; "--model"; "gpt-4"; "hello"] argv);

      test_case "codex runner via selector c" `Quick (fun () ->
        let parsed = Parser.parse_args ["c"; "hello"] in
        let argv, env, _warnings = Parser.resolve_command parsed None in
        check (list string) "argv" ["codex"; "hello"] argv;
        check (list (pair string string)) "env" [] env);

      test_case "codex runner via selector cx" `Quick (fun () ->
        let parsed = Parser.parse_args ["cx"; "hello"] in
        let argv, env, _warnings = Parser.resolve_command parsed None in
        check (list string) "argv" ["codex"; "hello"] argv;
        check (list (pair string string)) "env" [] env);

      test_case "provider env override" `Quick (fun () ->
        let parsed = Parser.parse_args [":anthropic:sonnet-4"; "hello"] in
        let argv, env, _warnings = Parser.resolve_command parsed None in
        check (list string) "argv" ["opencode"; "run"; "hello"] argv;
        check (list (pair string string)) "env"
          [("CCC_PROVIDER", "anthropic")] env);

      test_case "empty prompt raises" `Quick (fun () ->
        check_raises "empty" Parser.Empty_prompt (fun () ->
          let parsed = pa ~runner:None ~thinking:None ~provider:None
            ~model:None ~alias:None ~prompt:"" in
          ignore (Parser.resolve_command parsed None)));

      test_case "whitespace prompt raises" `Quick (fun () ->
        check_raises "ws" Parser.Empty_prompt (fun () ->
          let parsed = pa ~runner:None ~thinking:None ~provider:None
            ~model:None ~alias:None ~prompt:"   " in
          ignore (Parser.resolve_command parsed None)));

      test_case "preset with config" `Quick (fun () ->
        let parsed = Parser.parse_args ["@review"; "fix"; "bugs"] in
        let config = {
          Parser.default_runner = "oc"; Parser.default_provider = "";
          Parser.default_model = ""; Parser.default_thinking = None;
          Parser.aliases = [
            ("review", {
              Parser.ad_runner = Some "claude";
              Parser.ad_thinking = Some 3;
              Parser.ad_provider = None;
              Parser.ad_model = None;
              Parser.ad_agent = None;
            })
          ];
          Parser.abbreviations = [];
        } in
        let argv, env, _warnings = Parser.resolve_command parsed (Some config) in
        check (list string) "argv"
          ["claude"; "--thinking"; "enabled"; "--effort"; "high"; "fix bugs"] argv;
        check (list (pair string string)) "env" [] env);

      test_case "name without preset falls back to agent" `Quick (fun () ->
        let parsed = Parser.parse_args ["@reviewer"; "fix"; "bugs"] in
        let argv, env, warnings = Parser.resolve_command parsed None in
        check (list string) "argv"
          ["opencode"; "run"; "--agent"; "reviewer"; "fix bugs"] argv;
        check (list (pair string string)) "env" [] env;
        check (list string) "warnings" [] warnings);

      test_case "preset agent wins over name fallback" `Quick (fun () ->
        let parsed = Parser.parse_args ["@reviewer"; "fix"; "bugs"] in
        let config = {
          Parser.default_runner = "oc"; Parser.default_provider = "";
          Parser.default_model = ""; Parser.default_thinking = None;
          Parser.aliases = [
            ("reviewer", {
              Parser.ad_runner = None;
              Parser.ad_thinking = None;
              Parser.ad_provider = None;
              Parser.ad_model = None;
              Parser.ad_agent = Some "specialist";
            })
          ];
          Parser.abbreviations = [];
        } in
        let argv, env, warnings = Parser.resolve_command parsed (Some config) in
        check (list string) "argv"
          ["opencode"; "run"; "--agent"; "specialist"; "fix bugs"] argv;
        check (list (pair string string)) "env" [] env;
        check (list string) "warnings" [] warnings);

      test_case "unsupported runner warns and ignores agent" `Quick (fun () ->
        let parsed = Parser.parse_args ["rc"; "@reviewer"; "fix"; "bugs"] in
        let argv, env, warnings = Parser.resolve_command parsed None in
        check (list string) "argv" ["roocode"; "fix bugs"] argv;
        check (list (pair string string)) "env" [] env;
        check (list string) "warnings"
          ["warning: runner \"roocode\" does not support agents; ignoring @reviewer"]
          warnings);

      test_case "runner registry binaries and agent flags" `Quick (fun () ->
        let info name =
          match Hashtbl.find_opt Parser.runner_registry name with
          | Some info -> info
          | None -> failwith name
        in
        check string "opencode binary" "opencode" (info "opencode").Parser.binary;
        check string "claude binary" "claude" (info "claude").Parser.binary;
        check string "cc binary" "claude" (info "cc").Parser.binary;
        check string "c binary" "codex" (info "c").Parser.binary;
        check string "cx binary" "codex" (info "cx").Parser.binary;
        check string "codex binary" "codex" (info "codex").Parser.binary;
        check string "rc binary" "roocode" (info "rc").Parser.binary;
        check string "roocode binary" "roocode" (info "roocode").Parser.binary;
        check string "crush binary" "crush" (info "crush").Parser.binary;
        check string "opencode agent" "--agent" (info "opencode").Parser.agent_flag;
        check string "claude agent" "--agent" (info "claude").Parser.agent_flag;
        check string "kimi agent" "--agent" (info "kimi").Parser.agent_flag;
        check string "codex agent" "" (info "codex").Parser.agent_flag;
        check string "roocode agent" "" (info "roocode").Parser.agent_flag;
        check string "crush agent" "" (info "crush").Parser.agent_flag);

      test_case "kimi thinking +2" `Quick (fun () ->
        let parsed = Parser.parse_args ["kimi"; "+2"; "test"] in
        let argv, _, _warnings = Parser.resolve_command parsed None in
        check (list string) "argv" ["kimi"; "--thinking"; "test"] argv);

      test_case "crush ignores model flag" `Quick (fun () ->
        let parsed = Parser.parse_args ["crush"; ":model-1"; "hello"] in
        let argv, _, _warnings = Parser.resolve_command parsed None in
        check (list string) "argv" ["crush"; "hello"] argv);

      test_case "config defaults" `Quick (fun () ->
        let parsed = Parser.parse_args ["hello"] in
        let config = {
          Parser.default_runner = "claude";
          Parser.default_provider = "anthropic";
          Parser.default_model = "sonnet-4";
          Parser.default_thinking = Some 2;
          Parser.aliases = [];
          Parser.abbreviations = [];
        } in
        let argv, env, _warnings = Parser.resolve_command parsed (Some config) in
        check (list string) "argv"
          ["claude"; "--thinking"; "enabled"; "--effort"; "medium";
           "--model"; "sonnet-4"; "hello"] argv;
        check (list (pair string string)) "env"
          [("CCC_PROVIDER", "anthropic")] env);

      test_case "abbrev remaps runner selector cc" `Quick (fun () ->
        let parsed = Parser.parse_args ["cc"; "hello"] in
        let config = {
          Parser.default_runner = "oc"; Parser.default_provider = "";
          Parser.default_model = ""; Parser.default_thinking = None;
          Parser.aliases = [];
          Parser.abbreviations = [("cc", "codex")];
        } in
        let argv, _, _warnings = Parser.resolve_command parsed (Some config) in
        check (list string) "argv" ["codex"; "hello"] argv
      );

      test_case "abbrev to claude with thinking" `Quick (fun () ->
        let parsed = Parser.parse_args ["rc"; "+1"; "hello"] in
        let config = {
          Parser.default_runner = "oc"; Parser.default_provider = "";
          Parser.default_model = ""; Parser.default_thinking = None;
          Parser.aliases = [];
          Parser.abbreviations = [("rc", "claude")];
        } in
        let argv, _, _warnings = Parser.resolve_command parsed (Some config) in
        check (list string) "argv"
          ["claude"; "--thinking"; "enabled"; "--effort"; "low"; "hello"] argv);

      test_case "opencode thinking ignored (no flags)" `Quick (fun () ->
        let parsed = Parser.parse_args ["oc"; "+2"; "hello"] in
        let argv, _, _warnings = Parser.resolve_command parsed None in
        check (list string) "argv" ["opencode"; "run"; "hello"] argv);
    ]);
  ]

File: ruby/test/test_parser.rb
# frozen_string_literal: true

require "minitest/autorun"
require "call_coding_clis"

class TestParser < Minitest::Test
  def test_parse_prompt_only
    parsed = CallCodingClis::Parser.parse_args(["fix the tests"])
    assert_nil parsed.runner
    assert_nil parsed.thinking
    assert_nil parsed.provider
    assert_nil parsed.model
    assert_nil parsed.alias_name
    assert_equal "fix the tests", parsed.prompt
  end

  def test_parse_runner_selector
    parsed = CallCodingClis::Parser.parse_args(["claude", "hello"])
    assert_equal "claude", parsed.runner
    assert_equal "hello", parsed.prompt
  end

  def test_parse_thinking_flag
    parsed = CallCodingClis::Parser.parse_args(["+3", "do stuff"])
    assert_equal 3, parsed.thinking
    assert_equal "do stuff", parsed.prompt
  end

  def test_parse_provider_model
    parsed = CallCodingClis::Parser.parse_args([":anthropic:claude-3.5", "prompt"])
    assert_equal "anthropic", parsed.provider
    assert_equal "claude-3.5", parsed.model
    assert_equal "prompt", parsed.prompt
  end

  def test_parse_model_only
    parsed = CallCodingClis::Parser.parse_args([":gpt-4o", "prompt"])
    assert_nil parsed.provider
    assert_equal "gpt-4o", parsed.model
    assert_equal "prompt", parsed.prompt
  end

  def test_parse_alias
    parsed = CallCodingClis::Parser.parse_args(["@fast", "prompt"])
    assert_equal "fast", parsed.alias_name
    assert_equal "prompt", parsed.prompt
  end

  def test_parse_full_combo
    parsed = CallCodingClis::Parser.parse_args(["claude", "+2", ":anthropic:opus", "@fast", "fix it"])
    assert_equal "claude", parsed.runner
    assert_equal 2, parsed.thinking
    assert_equal "anthropic", parsed.provider
    assert_equal "opus", parsed.model
    assert_equal "fast", parsed.alias_name
    assert_equal "fix it", parsed.prompt
  end

  def test_parse_runner_abbreviation_oc
    parsed = CallCodingClis::Parser.parse_args(["oc", "prompt"])
    assert_equal "oc", parsed.runner
  end

  def test_parse_runner_abbreviation_cc
    parsed = CallCodingClis::Parser.parse_args(["cc", "prompt"])
    assert_equal "cc", parsed.runner
  end

  def test_parse_runner_abbreviation_c
    parsed = CallCodingClis::Parser.parse_args(["c", "prompt"])
    assert_equal "c", parsed.runner
  end

  def test_parse_runner_abbreviation_cx
    parsed = CallCodingClis::Parser.parse_args(["cx", "prompt"])
    assert_equal "cx", parsed.runner
  end

  def test_parse_runner_case_insensitive
    parsed = CallCodingClis::Parser.parse_args(["Claude", "prompt"])
    assert_equal "claude", parsed.runner
  end

  def test_parse_positional_after_non_positional
    parsed = CallCodingClis::Parser.parse_args(["+1", "fix", "tests"])
    assert_equal 1, parsed.thinking
    assert_equal "fix tests", parsed.prompt
  end

  def test_parse_runner_only_once
    parsed = CallCodingClis::Parser.parse_args(["claude", "oc", "prompt"])
    assert_equal "claude", parsed.runner
    assert_equal "oc prompt", parsed.prompt
  end

  def test_resolve_default_runner
    parsed = CallCodingClis::Parser.parse_args(["hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["opencode", "run", "hello"], argv
    assert_equal({}, env)
  end

  def test_resolve_claude_runner
    parsed = CallCodingClis::Parser.parse_args(["claude", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["claude", "hello"], argv
  end

  def test_resolve_thinking_flags_claude
    parsed = CallCodingClis::Parser.parse_args(["claude", "+2", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["claude", "--thinking", "enabled", "--effort", "medium", "hello"], argv
  end

  def test_resolve_thinking_flags_kimi
    parsed = CallCodingClis::Parser.parse_args(["kimi", "+0", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["kimi", "--no-thinking", "hello"], argv
  end

  def test_resolve_model_flag_claude
    parsed = CallCodingClis::Parser.parse_args(["claude", ":gpt-4o", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["claude", "--model", "gpt-4o", "hello"], argv
  end

  def test_resolve_model_flag_codex
    parsed = CallCodingClis::Parser.parse_args(["codex", ":gpt-4o", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["codex", "exec", "--model", "gpt-4o", "hello"], argv
  end

  def test_resolve_model_flag_codex_via_c
    parsed = CallCodingClis::Parser.parse_args(["c", ":gpt-4o", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["codex", "exec", "--model", "gpt-4o", "hello"], argv
  end

  def test_resolve_model_flag_codex_via_cx
    parsed = CallCodingClis::Parser.parse_args(["cx", ":gpt-4o", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["codex", "exec", "--model", "gpt-4o", "hello"], argv
  end

  def test_resolve_no_model_flag_opencode
    parsed = CallCodingClis::Parser.parse_args(["oc", ":gpt-4o", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["opencode", "run", "hello"], argv
  end

  def test_resolve_no_model_flag_crush
    parsed = CallCodingClis::Parser.parse_args(["crush", ":gpt-4o", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["crush", "hello"], argv
  end

  def test_resolve_no_model_flag_roocode
    parsed = CallCodingClis::Parser.parse_args(["rc", ":gpt-4o", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["roocode", "hello"], argv
  end

  def test_resolve_name_falls_back_to_agent_for_opencode
    parsed = CallCodingClis::Parser.parse_args(["@reviewer", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["opencode", "run", "--agent", "reviewer", "hello"], argv
    assert_equal({}, env)
  end

  def test_resolve_name_uses_agent_flag_for_claude
    parsed = CallCodingClis::Parser.parse_args(["claude", "@reviewer", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["claude", "--agent", "reviewer", "hello"], argv
    assert_equal({}, env)
  end

  def test_resolve_name_uses_agent_flag_for_kimi
    parsed = CallCodingClis::Parser.parse_args(["kimi", "@reviewer", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["kimi", "--agent", "reviewer", "hello"], argv
    assert_equal({}, env)
  end

  def test_resolve_provider_env
    parsed = CallCodingClis::Parser.parse_args([":anthropic:gpt-4o", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal "anthropic", env["CCC_PROVIDER"]
  end

  def test_resolve_empty_prompt_error
    parsed = CallCodingClis::Parser.parse_args([])
    assert_raises(ArgumentError) do
      CallCodingClis::Parser.resolve_command(parsed)
    end
  end

  def test_resolve_empty_prompt_error_message
    parsed = CallCodingClis::Parser.parse_args([])
    err = assert_raises(ArgumentError) do
      CallCodingClis::Parser.resolve_command(parsed)
    end
    assert_equal "prompt must not be empty", err.message
  end

  def test_resolve_config_defaults
    parsed = CallCodingClis::Parser.parse_args(["hello"])
    config = CallCodingClis::Parser::CccConfig.new(
      default_runner: "claude",
      default_provider: "anthropic",
      default_model: "opus",
      default_thinking: 3
    )
    argv, env = CallCodingClis::Parser.resolve_command(parsed, config)
    assert_equal "claude", argv[0]
    assert_includes argv, "--thinking"
    assert_includes argv, "enabled"
    assert_includes argv, "--effort"
    assert_includes argv, "high"
    assert_includes argv, "--model"
    assert_includes argv, "opus"
    assert_equal "anthropic", env["CCC_PROVIDER"]
  end

  def test_resolve_alias
    parsed = CallCodingClis::Parser.parse_args(["@fast", "hello"])
    config = CallCodingClis::Parser::CccConfig.new(
      aliases: {
        "fast" => CallCodingClis::Parser::AliasDef.new(
          runner: "claude",
          thinking: 4,
          provider: "anthropic",
          model: "opus"
        )
      }
    )
    argv, env = CallCodingClis::Parser.resolve_command(parsed, config)
    assert_equal "claude", argv[0]
    assert_includes argv, "--thinking"
    assert_includes argv, "enabled"
    assert_includes argv, "--effort"
    assert_includes argv, "max"
    assert_includes argv, "--model"
    assert_includes argv, "opus"
    assert_equal "anthropic", env["CCC_PROVIDER"]
    assert_equal "hello", argv.last
  end

  def test_resolve_alias_agent
    parsed = CallCodingClis::Parser.parse_args(["@fast", "hello"])
    config = CallCodingClis::Parser::CccConfig.new(
      aliases: {
        "fast" => CallCodingClis::Parser::AliasDef.new(
          agent: "specialist"
        )
      }
    )
    argv, env = CallCodingClis::Parser.resolve_command(parsed, config)
    assert_equal ["opencode", "run", "--agent", "specialist", "hello"], argv
    assert_equal({}, env)
  end

  def test_resolve_alias_runner_overridden_by_explicit
    parsed = CallCodingClis::Parser.parse_args(["kimi", "@fast", "hello"])
    config = CallCodingClis::Parser::CccConfig.new(
      aliases: {
        "fast" => CallCodingClis::Parser::AliasDef.new(runner: "claude")
      }
    )
    argv, _env = CallCodingClis::Parser.resolve_command(parsed, config)
    assert_equal "kimi", argv[0]
  end

  def test_resolve_unknown_name_falls_back_to_agent
    parsed = CallCodingClis::Parser.parse_args(["@nonexistent", "hello"])
    argv, env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal ["opencode", "run", "--agent", "nonexistent", "hello"], argv
    assert_equal({}, env)
  end

  def test_resolve_name_warning_for_unsupported_runner
    parsed = CallCodingClis::Parser.parse_args(["rc", "@reviewer", "hello"])
    warnings = []
    argv, env = CallCodingClis::Parser.resolve_command(parsed, nil, warnings: warnings)
    assert_equal ["roocode", "hello"], argv
    assert_equal({}, env)
    assert_equal ['warning: runner "rc" does not support agents; ignoring @reviewer'], warnings
  end

  def test_resolve_name_warning_for_codex_runner
    parsed = CallCodingClis::Parser.parse_args(["c", "@reviewer", "hello"])
    warnings = []
    argv, env = CallCodingClis::Parser.resolve_command(parsed, nil, warnings: warnings)
    assert_equal ["codex", "exec", "hello"], argv
    assert_equal({}, env)
    assert_equal ['warning: runner "c" does not support agents; ignoring @reviewer'], warnings
  end

  def test_resolve_abbreviation_via_config
    parsed = CallCodingClis::Parser.parse_args(["oc", "hello"])
    config = CallCodingClis::Parser::CccConfig.new(
      abbreviations: { "oc" => "claude" }
    )
    argv, _env = CallCodingClis::Parser.resolve_command(parsed, config)
    assert_equal "claude", argv[0]
  end

  def test_runner_registry_has_abbreviations
    assert_equal CallCodingClis::Parser::RUNNER_REGISTRY["oc"],
                 CallCodingClis::Parser::RUNNER_REGISTRY["opencode"]
    assert_equal CallCodingClis::Parser::RUNNER_REGISTRY["cc"],
                 CallCodingClis::Parser::RUNNER_REGISTRY["claude"]
    assert_equal CallCodingClis::Parser::RUNNER_REGISTRY["c"],
                 CallCodingClis::Parser::RUNNER_REGISTRY["codex"]
    assert_equal CallCodingClis::Parser::RUNNER_REGISTRY["cx"],
                 CallCodingClis::Parser::RUNNER_REGISTRY["codex"]
    assert_equal CallCodingClis::Parser::RUNNER_REGISTRY["k"],
                 CallCodingClis::Parser::RUNNER_REGISTRY["kimi"]
    assert_equal CallCodingClis::Parser::RUNNER_REGISTRY["rc"],
                 CallCodingClis::Parser::RUNNER_REGISTRY["roocode"]
    assert_equal CallCodingClis::Parser::RUNNER_REGISTRY["cr"],
                 CallCodingClis::Parser::RUNNER_REGISTRY["crush"]
  end

  def test_config_struct_defaults
    config = CallCodingClis::Parser::CccConfig.new
    assert_equal "oc", config.default_runner
    assert_equal "", config.default_provider
    assert_equal "", config.default_model
    assert_nil config.default_thinking
    assert_equal({}, config.aliases)
    assert_equal({}, config.abbreviations)
  end

  def test_parse_multiple_words_as_prompt
    parsed = CallCodingClis::Parser.parse_args(["fix", "the", "tests"])
    assert_equal "fix the tests", parsed.prompt
  end

  def test_parse_thinking_zero
    parsed = CallCodingClis::Parser.parse_args(["+0", "hello"])
    assert_equal 0, parsed.thinking
  end

  def test_parse_thinking_four
    parsed = CallCodingClis::Parser.parse_args(["+4", "hello"])
    assert_equal 4, parsed.thinking
  end

  def test_parse_thinking_rejects_five
    parsed = CallCodingClis::Parser.parse_args(["+5", "hello"])
    assert_nil parsed.thinking
    assert_equal "+5 hello", parsed.prompt
  end

  def test_resolve_whitespaced_prompt_stripped
    parsed = CallCodingClis::Parser.parse_args(["  hello  "])
    argv, _env = CallCodingClis::Parser.resolve_command(parsed)
    assert_equal "hello", argv.last
  end
end

class TestConfig < Minitest::Test
  def test_load_config_missing_file
    config = CallCodingClis::Config.load_config("/nonexistent/path/config.toml")
    assert_equal "oc", config.default_runner
    assert_equal({}, config.aliases)
  end

  def test_load_config_from_file
    require "tempfile"
    file = Tempfile.new(["ccc_test", ".toml"])
    file.write(<<~TOML)
      [defaults]
      runner = "claude"
      provider = "anthropic"
      model = "opus"
      thinking = 3

      [abbreviations]
      my = "claude"

      [aliases.fast]
      runner = "claude"
      thinking = 4
      provider = "anthropic"
      model = "opus"
      agent = "specialist"
    TOML
    file.close

    config = CallCodingClis::Config.load_config(file.path)
    assert_equal "claude", config.default_runner
    assert_equal "anthropic", config.default_provider
    assert_equal "opus", config.default_model
    assert_equal 3, config.default_thinking
    assert_equal({ "my" => "claude" }, config.abbreviations)
    assert_equal "claude", config.aliases["fast"].runner
    assert_equal 4, config.aliases["fast"].thinking
    assert_equal "anthropic", config.aliases["fast"].provider
    assert_equal "opus", config.aliases["fast"].model
    assert_equal "specialist", config.aliases["fast"].agent
  ensure
    file.unlink
  end

  def test_load_config_partial
    require "tempfile"
    file = Tempfile.new(["ccc_test", ".toml"])
    file.write(<<~TOML)
      [defaults]
      runner = "kimi"
    TOML
    file.close

    config = CallCodingClis::Config.load_config(file.path)
    assert_equal "kimi", config.default_runner
    assert_equal "", config.default_provider
  ensure
    file.unlink
  end
end
